### ISSUE TITLE
The training time and can this code use apex for faster training?

### ISSUE THREAD

**Issue by cientgu** [rosinality/stylegan2-pytorch#90]
I train for 800000 iterations on 256*256 resolution, batchsize sets to 32 on 4xV100, the gpu-utilization is almost 100%, each iteration needs about 4 seconds, all the training need 900 hours, it's slow than the authors which use almost the same setting (https://github.com/rosinality/stylegan2-pytorch/issues/2#issuecomment-583703060_), i don't know why, could you help me? Another thing is could this code try on apex to faster the training?

**Comment by rosinality** [comment]
It's strange. Did you set batch size to 8 so total batch sizes to be 32 with 4 gpus?

**Comment by cientgu** [comment]
Thank you for your reply.
Actually, I set batch for 32 in the training command and use 4 gpus, so actually i use the batchsize of 128? Is that right? Thank you for point it out.  And when i set the batch to 32, the gpu-memory cost is about 30G for each GPU, so you mean i need to set it to 8 when i use 4 gpus and the memory of each will decrease?
Another thing about apex is: when i set to O1(register both upfirdn2d and fused_act), the training time increase about 5 times compared with O0..... I am not sure why.
Thank you.

**Comment by rosinality** [comment]
Yes, batch size parameter corresponds to batch size per gpu. Sorry for confusion. (I always wonder about batch per gpu vs total batch sizes.) If you decrease batch sizes then training speed & memory consumption will be quite improved.

About amp current hard to know why. I think you can use O2 or O3 and if performance is improved, then automatic conversion to fp32 was problematic. If not, you can check performances of fp16 operations individually.

**Comment by cientgu** [comment]
Thank you for your relpy. There are something wrong here for O2 and O3, i will try it in the next few days.

**Comment by songyoyo** [comment]
> Thank you for your relpy. There are something wrong here for O2 and O3, i will try it in the next few days.

When I use O1, there are something wrong. What should I do to solve this problem?
![image](https://user-images.githubusercontent.com/42490856/89983700-a3454b80-dcaa-11ea-992f-c69fac04bb4a.png)

**Comment by rosinality** [comment]
You can manually cast input or kernel of upfirdn2d to fp16 or fp32.

**Comment by cientgu** [comment]
> > Thank you for your relpy. There are something wrong here for O2 and O3, i will try it in the next few days.
> 
> When I use O1, there are something wrong. What should I do to solve this problem?
> ![image](https://user-images.githubusercontent.com/42490856/89983700-a3454b80-dcaa-11ea-992f-c69fac04bb4a.png)

Have you successfully use apex in this code for faster training or saving memory? I set upfirdn2d and fused_act to fp16, and follow the apex.amp example ( https://github.com/NVIDIA/apex/blob/master/examples/dcgan/main_amp.py ), but the O1 is still slower and cost larger memory, have you solved it?

**Comment by cientgu** [comment]
When I try this code in PyTorch1.6 with the "torch.cuda.amp", it's still slower than original setting ( I set upfirdn2d and fused_act to FP32 and others use the default setting), it's so strange, does anyboby know why?

**Comment by cientgu** [comment]
> When I try this code in PyTorch1.6 with the "torch.cuda.amp", it's still slower than original setting ( I set upfirdn2d and fused_act to FP32 and others use the default setting), it's so strange, does anyboby know why?

When I set both upfirdn2d and fused_act to FP16, It's still slowly and the GPU memory increase compared with FP32, I don't know why it happens.
